File: src/laohuangcode/bash_runner.py
"""Streaming, cancellable execution for the Bash tool.

The runner deliberately owns process lifecycle and pipe draining, while callers
decide how emitted events are presented.  It targets macOS and Linux, where a
new process session lets cancellation terminate the whole command process group.
"""

from __future__ import annotations

import codecs
from dataclasses import dataclass
import os
from pathlib import Path
import queue
import signal
import subprocess
import threading
import time
from typing import Any, Literal, Mapping


BashStatus = Literal[
    "completed",
    "failed",
    "timed_out",
    "cancelled",
    "spawn_failed",
]

_READ_SIZE = 4096
_FLUSH_INTERVAL = 0.04
_FLUSH_CHARS = 4096
_TERMINATION_GRACE = 2.0
MODEL_API_KEY_ENV_NAMES = (
    "OPENAI_API_KEY",
    "DEEPSEEK_API_KEY",
    "LAOHUANG_API_KEY",
)


@dataclass(slots=True)
class ToolExecutionContext:
    """Runtime metadata and optional cooperative services for a tool call."""

    session_id: str | None = None
    task_id: str | None = None
    tool_call_id: str | None = None
    cancel_token: Any | None = None
    event_sink: Any | None = None

    def is_cancelled(self) -> bool:
        token = self.cancel_token
        if token is None:
            return False
        checker = getattr(token, "is_cancelled", None)
        if callable(checker):
            return bool(checker())
        checker = getattr(token, "is_set", None)
        if callable(checker):
            return bool(checker())
        return bool(getattr(token, "cancelled", False))

    @property
    def cancellation_reason(self) -> str:
        reason = getattr(self.cancel_token, "reason", None)
        return str(reason) if reason else "cancelled"

    def publish(self, kind: str, payload: dict[str, Any]) -> None:
        """Publish through EventBus, with a tiny callback fallback for tests."""

        sink = self.event_sink
        if sink is None:
            return

        publisher = getattr(sink, "publish", None)
        if callable(publisher):
            # Import lazily so the runner remains usable during runtime startup
            # and does not create a dependency cycle with the event definitions.
            from .events import EventKind, EventSource

            publisher(
                EventKind(kind),
                source=EventSource.TOOL,
                session_id=self.session_id or "local",
                task_id=self.task_id,
                correlation_id=self.tool_call_id,
                payload=payload,
            )
            return

        if callable(sink):
            sink(kind, payload)


@dataclass(slots=True)
class BashResult:
    status: BashStatus
    stdout: str = ""
    stderr: str = ""
    exit_code: int | None = None
    error: str | None = None
    duration_ms: int = 0
    truncated: bool = False

    @property
    def ok(self) -> bool:
        return self.status == "completed" and self.exit_code == 0

    def as_dict(self) -> dict[str, Any]:
        result: dict[str, Any] = {
            "ok": self.ok,
            "status": self.status,
            "exit_code": self.exit_code,
            "stdout": self.stdout,
            "stderr": self.stderr,
            "duration_ms": self.duration_ms,
            "truncated": self.truncated,
        }
        if self.error:
            result["error"] = self.error
        return result


class _BoundedOutput:
    """Keep a 40% head and 60% tail without retaining unbounded output."""

    def __init__(self, limit: int) -> None:
        self.limit = max(0, limit)
        self.head_limit = int(self.limit * 0.4)
        self.tail_limit = self.limit - self.head_limit
        self.total_chars = 0
        self._whole = ""
        self._head = ""
        self._tail = ""
        self._truncated = False

    def append(self, text: str) -> None:
        if not text:
            return
        self.total_chars += len(text)
        if self.limit == 0:
            self._truncated = True
            return

        if not self._truncated and len(self._whole) + len(text) <= self.limit:
            self._whole += text
            return

        if not self._truncated:
            combined = self._whole + text
            self._head = combined[: self.head_limit]
            self._tail = combined[-self.tail_limit :] if self.tail_limit else ""
            self._whole = ""
            self._truncated = True
            return

        if self.tail_limit:
            self._tail = (self._tail + text)[-self.tail_limit :]

    def render(self) -> str:
        if not self._truncated:
            return self._whole
        retained = len(self._head) + len(self._tail)
        omitted = max(0, self.total_chars - retained)
        marker = f"\n...[truncated {omitted} chars]...\n"
        return self._head + marker + self._tail

    @property
    def truncated(self) -> bool:
        return self._truncated


class _TerminalSanitizer:
    """Incrementally remove terminal control sequences from untrusted output."""

    def __init__(self) -> None:
        self._state = "normal"

    def feed(self, text: str) -> str:
        safe: list[str] = []
        for character in text:
            code = ord(character)
            if self._state == "normal":
                if character == "\x1b":
                    self._state = "escape"
                elif character == "\r":
                    safe.append("\n")
                elif character in ("\n", "\t") or code >= 0x20:
                    if code != 0x7F:
                        safe.append(character)
                continue

            if self._state == "escape":
                if character == "[":
                    self._state = "csi"
                elif character in ("P", "X", "^", "_"):
                    self._state = "string"
                elif character == "]":
                    self._state = "osc"
                else:
                    self._state = "normal"
                continue

            if self._state == "csi":
                if 0x40 <= code <= 0x7E:
                    self._state = "normal"
                continue

            if self._state in ("osc", "string"):
                if character == "\x07":
                    self._state = "normal"
                elif character == "\x1b":
                    self._state = "string_escape"
                continue

            if self._state == "string_escape":
                if character == "\\":
                    self._state = "normal"
                elif character != "\x1b":
                    self._state = "string"

        return "".join(safe)


@dataclass(slots=True)
class _ReaderMessage:
    stream: Literal["stdout", "stderr"]
    text: str | None


def _read_pipe(
    stream: Literal["stdout", "stderr"],
    pipe: Any,
    messages: queue.Queue[_ReaderMessage],
) -> None:
    decoder = codecs.getincrementaldecoder("utf-8")(errors="replace")
    sanitizer = _TerminalSanitizer()
    try:
        while True:
            data = os.read(pipe.fileno(), _READ_SIZE)
            if not data:
                break
            text = sanitizer.feed(decoder.decode(data, final=False))
            if text:
                messages.put(_ReaderMessage(stream, text))
        final_text = sanitizer.feed(decoder.decode(b"", final=True))
        if final_text:
            messages.put(_ReaderMessage(stream, final_text))
    except (OSError, ValueError):
        # Pipes can be closed by cancellation cleanup while readers are active.
        pass
    finally:
        messages.put(_ReaderMessage(stream, None))


def _terminate_process_group(
    process: subprocess.Popen[bytes], grace_seconds: float = _TERMINATION_GRACE
) -> None:
    try:
        os.killpg(process.pid, signal.SIGTERM)
    except ProcessLookupError:
        return
    except OSError:
        process.terminate()

    deadline = time.monotonic() + grace_seconds
    while time.monotonic() < deadline:
        # poll() reaps an exited group leader; otherwise a zombie can make the
        # group look alive for the entire grace period.
        process.poll()
        try:
            os.killpg(process.pid, 0)
        except ProcessLookupError:
            process.wait(timeout=0)
            return
        except PermissionError:
            break
        time.sleep(0.02)

    try:
        os.killpg(process.pid, signal.SIGKILL)
    except ProcessLookupError:
        return
    except OSError:
        process.kill()
    try:
        process.wait(timeout=1.0)
    except subprocess.TimeoutExpired:
        pass


def _clean_environment(environment: Mapping[str, str]) -> dict[str, str]:
    cleaned = dict(environment)
    for name in MODEL_API_KEY_ENV_NAMES:
        cleaned.pop(name, None)
    return cleaned


def run_bash(
    command: str,
    *,
    cwd: Path,
    timeout: float,
    max_output_chars: int,
    context: ToolExecutionContext | None = None,
    env: Mapping[str, str] | None = None,
) -> BashResult:
    """Run one non-interactive Bash command and emit sanitized output deltas."""

    invoked_at = time.monotonic()
    execution = context or ToolExecutionContext()
    if execution.is_cancelled():
        result = BashResult(
            status="cancelled",
            error=execution.cancellation_reason,
            duration_ms=int((time.monotonic() - invoked_at) * 1000),
        )
        execution.publish("tool.finished", result.as_dict())
        return result

    environment = _clean_environment(env if env is not None else os.environ)
    try:
        process = subprocess.Popen(
            ["/bin/bash", "-lc", command],
            cwd=cwd,
            env=environment,
            stdin=subprocess.DEVNULL,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=False,
            start_new_session=True,
        )
    except (OSError, TypeError, ValueError) as error:
        result = BashResult(
            status="spawn_failed",
            error=str(error),
            duration_ms=int((time.monotonic() - invoked_at) * 1000),
        )
        execution.publish("tool.finished", result.as_dict())
        return result

    execution.publish(
        "tool.started",
        {"name": "bash", "arguments": {"command": command}},
    )
    assert process.stdout is not None
    assert process.stderr is not None

    messages: queue.Queue[_ReaderMessage] = queue.Queue(maxsize=256)
    readers = [
        threading.Thread(
            target=_read_pipe,
            args=("stdout", process.stdout, messages),
            name=f"bash-stdout-{process.pid}",
            daemon=True,
        ),
        threading.Thread(
            target=_read_pipe,
            args=("stderr", process.stderr, messages),
            name=f"bash-stderr-{process.pid}",
            daemon=True,
        ),
    ]
    for reader in readers:
        reader.start()

    outputs = {
        "stdout": _BoundedOutput(max_output_chars),
        "stderr": _BoundedOutput(max_output_chars),
    }
    pending = {"stdout": "", "stderr": ""}
    stream_sequence = {"stdout": 0, "stderr": 0}
    finished_streams: set[str] = set()
    started_at = time.monotonic()
    last_flush = started_at
    terminal_status: BashStatus | None = None

    def flush(stream: Literal["stdout", "stderr"]) -> None:
        text = pending[stream]
        if not text:
            return
        pending[stream] = ""
        outputs[stream].append(text)
        stream_sequence[stream] += 1
        execution.publish(
            "tool.output_delta",
            {
                "name": "bash",
                "stream": stream,
                "text": text,
                "stream_sequence": stream_sequence[stream],
            },
        )

    try:
        while len(finished_streams) < 2 or process.poll() is None:
            now = time.monotonic()
            if terminal_status is None and execution.is_cancelled():
                terminal_status = "cancelled"
                _terminate_process_group(process)
            elif terminal_status is None and timeout >= 0 and now - started_at >= timeout:
                terminal_status = "timed_out"
                _terminate_process_group(process)

            try:
                message = messages.get(timeout=0.01)
            except queue.Empty:
                message = None

            if message is not None:
                if message.text is None:
                    finished_streams.add(message.stream)
                else:
                    pending[message.stream] += message.text
                    if len(pending[message.stream]) >= _FLUSH_CHARS:
                        flush(message.stream)

            now = time.monotonic()
            if now - last_flush >= _FLUSH_INTERVAL:
                flush("stdout")
                flush("stderr")
                last_flush = now

        flush("stdout")
        flush("stderr")
    finally:
        if process.poll() is None:
            _terminate_process_group(process)
        for pipe in (process.stdout, process.stderr):
            try:
                pipe.close()
            except OSError:
                pass
        for reader in readers:
            reader.join(timeout=1.0)

    exit_code = process.poll()
    stdout = outputs["stdout"].render()
    stderr = outputs["stderr"].render()
    duration_ms = int((time.monotonic() - invoked_at) * 1000)
    truncated = outputs["stdout"].truncated or outputs["stderr"].truncated
    if terminal_status == "cancelled":
        result = BashResult(
            status="cancelled",
            stdout=stdout,
            stderr=stderr,
            exit_code=exit_code,
            error=execution.cancellation_reason,
            duration_ms=duration_ms,
            truncated=truncated,
        )
    elif terminal_status == "timed_out":
        result = BashResult(
            status="timed_out",
            stdout=stdout,
            stderr=stderr,
            exit_code=exit_code,
            error=f"Bash command timed out after {timeout} seconds",
            duration_ms=duration_ms,
            truncated=truncated,
        )
    elif exit_code == 0:
        result = BashResult(
            status="completed",
            stdout=stdout,
            stderr=stderr,
            exit_code=exit_code,
            duration_ms=duration_ms,
            truncated=truncated,
        )
    else:
        result = BashResult(
            status="failed",
            stdout=stdout,
            stderr=stderr,
            exit_code=exit_code,
            error=f"Bash command exited with code {exit_code}",
            duration_ms=duration_ms,
            truncated=truncated,
        )

    execution.publish("tool.finished", result.as_dict())
    return result

File: packages/shell/bash-local/src/bash-runner.ts
/**
 * Streaming, cancellable execution for the Bash tool.
 *
 * The runner deliberately owns process lifecycle and pipe draining, while
 * callers decide how emitted events are presented. Unix uses process-group
 * signals for cancellation; Windows uses taskkill to terminate the tree.
 */

import { execFile, spawn, type ChildProcess } from "node:child_process";
import { constants as osConstants } from "node:os";
import { win32 } from "node:path";
import { StringDecoder } from "node:string_decoder";

import type { CancelToken } from "@laohuang/runtime-protocol";
import { resolveBashPath } from "./bash-path.ts";
import { BashOutput, type BashOutputOptions } from "./bash-output.ts";

export type BashStatus =
  | "completed"
  | "failed"
  | "timed_out"
  | "cancelled"
  | "spawn_failed";

type StreamName = "stdout" | "stderr";

const FLUSH_INTERVAL_MS = 100;
const TERMINATION_GRACE_MS = 2000;
const WATCHDOG_INTERVAL_MS = 10;
const POST_EXIT_IDLE_MS = 100;
const TERMINATION_EXIT_GRACE_MS = 1000;
const TASKKILL_TIMEOUT_MS = 5000;

export const MODEL_API_KEY_ENV_NAMES = [
  "OPENAI_API_KEY",
  "DEEPSEEK_API_KEY",
  "LAOHUANG_API_KEY",
] as const;

/** Structural minimum for the CancelToken owned by runtime-protocol. */
export interface CancelTokenLike {
  isCancelled?: CancelToken["isCancelled"];
  isSet?: () => boolean;
  cancelled?: boolean;
  reason?: unknown;
}

/** Structural minimum for the EventBus owned by runtime-protocol. */
export interface ToolEventPublisher {
  publish(
    kind: string,
    options: {
      source: string;
      session_id: string;
      task_id: string | null;
      correlation_id: string | null;
      payload: Record<string, unknown>;
    },
  ): unknown;
}

export type ToolEventSink =
  | ToolEventPublisher
  | ((kind: string, payload: Record<string, unknown>) => unknown);

export interface ToolExecutionContextInit {
  sessionId?: string | null;
  taskId?: string | null;
  toolCallId?: string | null;
  cancelToken?: CancelTokenLike | null;
  eventSink?: ToolEventSink | null;
}

/** Runtime metadata and optional cooperative services for a bash command. */
export class ToolExecutionContext {
  sessionId: string | null;
  taskId: string | null;
  toolCallId: string | null;
  cancelToken: CancelTokenLike | null;
  eventSink: ToolEventSink | null;

  constructor(init: ToolExecutionContextInit = {}) {
    this.sessionId = init.sessionId ?? null;
    this.taskId = init.taskId ?? null;
    this.toolCallId = init.toolCallId ?? null;
    this.cancelToken = init.cancelToken ?? null;
    this.eventSink = init.eventSink ?? null;
  }

  isCancelled(): boolean {
    const token = this.cancelToken;
    if (token == null) return false;
    if (typeof token.isCancelled === "function") {
      return Boolean(token.isCancelled.call(token));
    }
    if (typeof token.isSet === "function") {
      return Boolean(token.isSet.call(token));
    }
    return Boolean(token.cancelled);
  }

  get cancellationReason(): string {
    const reason = this.cancelToken?.reason;
    return reason ? String(reason) : "cancelled";
  }

  /** Publish through an EventBus, with a tiny callback fallback for tests. */
  publish(kind: string, payload: Record<string, unknown>): unknown {
    const sink = this.eventSink;
    if (sink == null) return;

    if (typeof sink === "function") {
      return sink(kind, payload);
    }

    if (typeof sink.publish === "function") {
      return sink.publish(kind, {
        source: "tool",
        session_id: this.sessionId ?? "local",
        task_id: this.taskId,
        correlation_id: this.toolCallId,
        payload,
      });
    }
  }
}

export interface BashResultInit {
  status: BashStatus;
  stdout?: string;
  stderr?: string;
  exitCode?: number | null;
  error?: string | null;
  durationMs?: number;
  truncated?: boolean;
  outputComplete?: boolean;
  outputFiles?: Record<StreamName, string> | null;
  outputFileComplete?: boolean;
  outputFileError?: string | null;
  stdoutStartMidLine?: boolean;
  stderrStartMidLine?: boolean;
}

export class BashResult {
  status: BashStatus;
  stdout: string;
  stderr: string;
  exitCode: number | null;
  error: string | null;
  durationMs: number;
  truncated: boolean;
  outputComplete: boolean;
  outputFiles: Record<StreamName, string> | null;
  outputFileComplete: boolean;
  outputFileError: string | null;
  stdoutStartMidLine: boolean;
  stderrStartMidLine: boolean;

  constructor(init: BashResultInit) {
    this.status = init.status;
    this.stdout = init.stdout ?? "";
    this.stderr = init.stderr ?? "";
    this.exitCode = init.exitCode ?? null;
    this.error = init.error ?? null;
    this.durationMs = init.durationMs ?? 0;
    this.truncated = init.truncated ?? false;
    this.outputComplete = init.outputComplete ?? true;
    this.outputFiles = init.outputFiles ?? null;
    this.outputFileComplete = init.outputFileComplete ?? false;
    this.outputFileError = init.outputFileError ?? null;
    this.stdoutStartMidLine = init.stdoutStartMidLine ?? false;
    this.stderrStartMidLine = init.stderrStartMidLine ?? false;
  }

  get ok(): boolean {
    return this.status === "completed" && this.exitCode === 0;
  }

  asDict(): Record<string, unknown> {
    const result: Record<string, unknown> = {
      ok: this.ok,
      status: this.status,
      exit_code: this.exitCode,
      stdout: this.stdout,
      stderr: this.stderr,
      duration_ms: this.durationMs,
      truncated: this.truncated,
      output_complete: this.outputComplete,
      output_files: this.outputFiles,
      output_file_complete: this.outputFileComplete,
      stdout_start_mid_line: this.stdoutStartMidLine,
      stderr_start_mid_line: this.stderrStartMidLine,
    };
    if (this.outputFileError) result["output_file_error"] = this.outputFileError;
    if (this.error) {
      result["error"] = this.error;
    }
    return result;
  }
}

type SanitizerState =
  | "normal"
  | "escape"
  | "csi"
  | "osc"
  | "string"
  | "stringEscape";

/** Incrementally remove terminal control sequences from untrusted output. */
class TerminalSanitizer {
  private state: SanitizerState = "normal";

  feed(text: string): string {
    let safe = "";
    for (const character of text) {
      const code = character.codePointAt(0) ?? 0;
      if (this.state === "normal") {
        if (character === "\x1b") {
          this.state = "escape";
        } else if (character === "\r") {
          safe += "\n";
        } else if (character === "\n" || character === "\t" || code >= 0x20) {
          if (code !== 0x7f) safe += character;
        }
        continue;
      }

      if (this.state === "escape") {
        if (character === "[") {
          this.state = "csi";
        } else if (
          character === "P" ||
          character === "X" ||
          character === "^" ||
          character === "_"
        ) {
          this.state = "string";
        } else if (character === "]") {
          this.state = "osc";
        } else {
          this.state = "normal";
        }
        continue;
      }

      if (this.state === "csi") {
        if (code >= 0x40 && code <= 0x7e) this.state = "normal";
        continue;
      }

      if (this.state === "osc" || this.state === "string") {
        if (character === "\x07") {
          this.state = "normal";
        } else if (character === "\x1b") {
          this.state = "stringEscape";
        }
        continue;
      }

      if (this.state === "stringEscape") {
        if (character === "\\") {
          this.state = "normal";
        } else if (character !== "\x1b") {
          this.state = "string";
        }
      }
    }
    return safe;
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function errnoCode(error: unknown): string | undefined {
  return (error as NodeJS.ErrnoException | undefined)?.code;
}

/**
 * Terminate the Windows process tree, or signal the Unix process group with
 * SIGTERM and escalate to SIGKILL after a grace period.
 */
async function terminateProcessTree(
  child: ChildProcess,
  graceMs: number = TERMINATION_GRACE_MS,
): Promise<void> {
  const pid = child.pid;
  if (pid === undefined) return;

  if (process.platform === "win32") {
    const taskkill = win32.join(process.env.SystemRoot ?? "C:\\Windows", "System32", "taskkill.exe");
    await new Promise<void>((resolve, reject) => {
      execFile(taskkill, ["/F", "/T", "/PID", String(pid)], {
        windowsHide: true,
        timeout: TASKKILL_TIMEOUT_MS,
        killSignal: "SIGKILL",
      }, (error) => {
        if (error) reject(new Error(`taskkill failed: ${error.message}`));
        else resolve();
      });
    });
    return;
  }

  try {
    process.kill(-pid, "SIGTERM");
  } catch (error) {
    if (errnoCode(error) === "ESRCH") return;
    try {
      child.kill("SIGTERM");
    } catch {
      // The leader may already be gone; the liveness loop below handles it.
    }
  }

  const deadline = Date.now() + graceMs;
  while (Date.now() < deadline) {
    try {
      process.kill(-pid, 0);
    } catch (error) {
      const code = errnoCode(error);
      if (code === "ESRCH") return;
      if (code === "EPERM") break;
    }
    await sleep(20);
  }

  try {
    process.kill(-pid, "SIGKILL");
  } catch (error) {
    if (errnoCode(error) === "ESRCH") return;
    try {
      child.kill("SIGKILL");
    } catch {
      // Nothing left to kill.
    }
  }
}

function cleanEnvironment(
  environment: Record<string, string | undefined>,
): Record<string, string | undefined> {
  const cleaned = { ...environment };
  for (const name of MODEL_API_KEY_ENV_NAMES) {
    delete cleaned[name];
  }
  return cleaned;
}

function toExitCode(
  code: number | null,
  signal: NodeJS.Signals | null,
): number | null {
  if (code !== null) return code;
  if (!signal) return null;
  const signalNumber: number | undefined = osConstants.signals[signal];
  return typeof signalNumber === "number" ? -signalNumber : null;
}

export interface RunBashOptions extends BashOutputOptions {
  cwd: string;
  shellPath?: string | undefined;
  /** Seconds; a negative value disables the timeout. */
  timeout: number;
  context?: Pick<ToolExecutionContext, "isCancelled" | "cancellationReason" | "publish"> | null;
  env?: Record<string, string | undefined> | null;
}

/** Run one non-interactive Bash command and emit bounded output snapshots. */
export async function runBash(
  command: string,
  options: RunBashOptions,
): Promise<BashResult> {
  const invokedAt = performance.now();
  const execution = options.context ?? new ToolExecutionContext();
  const output = new BashOutput(options);
  let publicationError: string | null = null;
  const publish = async (kind: string, payload: Record<string, unknown>): Promise<void> => {
    try {
      await execution.publish(kind, payload);
    } catch (error) {
      publicationError ??= `Bash event publication failed: ${String(error)}`;
    }
  };
  const finish = async (result: BashResult): Promise<BashResult> => {
    await publish("tool.finished", result.asDict());
    if (publicationError !== null) {
      result.status = "failed";
      if (!result.error?.includes(publicationError)) {
        result.error = [result.error, publicationError].filter(Boolean).join("; ");
      }
    }
    return result;
  };

  if (execution.isCancelled()) {
    const result = new BashResult({
      status: "cancelled",
      error: execution.cancellationReason,
      durationMs: Math.trunc(performance.now() - invokedAt),
    });
    return await finish(result);
  }

  const environment = cleanEnvironment(options.env ?? process.env);
  let child: ChildProcess;
  try {
    const shellPath = resolveBashPath({ shellPath: options.shellPath, env: environment });
    child = spawn(shellPath, ["-lc", command], {
      cwd: options.cwd,
      env: environment,
      stdio: ["ignore", "pipe", "pipe"],
      detached: process.platform !== "win32",
      windowsHide: true,
    });
  } catch (error) {
    const result = new BashResult({
      status: "spawn_failed",
      error: String(error),
      durationMs: Math.trunc(performance.now() - invokedAt),
    });
    return await finish(result);
  }
  const spawnError = await new Promise<Error | null>((resolve) => {
    child.once("spawn", () => resolve(null));
    child.once("error", (error) => resolve(error));
  });
  if (spawnError) {
    const result = new BashResult({
      status: "spawn_failed",
      error: String(spawnError),
      durationMs: Math.trunc(performance.now() - invokedAt),
    });
    return await finish(result);
  }
  // Later asynchronous kill failures must not surface as unhandled errors.
  child.on("error", () => {});

  const started = publish("tool.started", { name: "bash", arguments: { command } });
  let publishing: Promise<void> | null = null;
  let publishedRevision = 0;
  let exited = false;
  let lastOutputAt = performance.now();
  let outputIncomplete = false;
  const ingesting: Record<StreamName, boolean> = { stdout: false, stderr: false };

  const closeOutput = (force = false): void => {
    for (const stream of ["stdout", "stderr"] as const) {
      const pipe = child[stream];
      // Waiting for disk is backpressure, not an idle inherited pipe. Finish
      // persisting the current read and draining Node's buffered bytes first.
      if (!force && (ingesting[stream] || (pipe?.readableLength ?? 0) > 0)) continue;
      if (pipe && !pipe.readableEnded && !pipe.destroyed) {
        outputIncomplete = true;
        pipe.destroy();
      }
    }
  };

  const flush = (): void => {
    if (publishing !== null || publicationError !== null || publishedRevision === output.revision) return;
    const snapshot = output.snapshot();
    publishedRevision = output.revision;
    const revision = publishedRevision;
    // Only one publication is in flight. Further reads update bounded state;
    // the next tick sends the latest state instead of queueing every revision.
    publishing = (async () => {
      await started;
      for (const stream of ["stdout", "stderr"] as const) {
        if (publicationError !== null) break;
        await publish("tool.output_snapshot", {
          name: "bash", stream, text: snapshot[stream].text,
          stream_sequence: revision, truncated: snapshot[stream].truncated,
          start_mid_line: snapshot[stream].startMidLine,
        });
      }
    })().finally(() => { publishing = null; });
  };

  const readPipe = async (
    stream: StreamName,
    pipe: NodeJS.ReadableStream,
  ): Promise<void> => {
    const decoder = new StringDecoder("utf8");
    const sanitizer = new TerminalSanitizer();
    try {
      for await (const chunk of pipe) {
        lastOutputAt = performance.now();
        ingesting[stream] = true;
        try {
          await output.append(stream, sanitizer.feed(decoder.write(chunk as Buffer)));
        } finally {
          ingesting[stream] = false;
          lastOutputAt = performance.now();
        }
      }
    } catch {
      // Cancellation can tear down a pipe; never claim its output is complete.
      outputIncomplete = true;
    } finally {
      await output.append(stream, sanitizer.feed(decoder.end()));
    }
  };

  const stdoutDone = readPipe("stdout", child.stdout!);
  const stderrDone = readPipe("stderr", child.stderr!);
  type ExitInfo = {
    code: number | null;
    signal: NodeJS.Signals | null;
  };
  let resolveExit!: (info: ExitInfo) => void;
  const exitInfo = new Promise<ExitInfo>((resolve) => {
    resolveExit = resolve;
  });
  const onExit = (code: number | null, signal: NodeJS.Signals | null): void => {
    exited = true;
    lastOutputAt = performance.now();
    resolveExit({ code, signal });
  };
  child.once("exit", onExit);

  let terminalStatus: BashStatus | null = null;
  const startedAt = performance.now();
  let lastFlush = startedAt;
  const timeoutMs = options.timeout * 1000;
  let termination: Promise<void> | null = null;
  let terminationCompletedAt: number | null = null;
  let cleanupError: string | null = null;
  const terminate = (): void => {
    if (termination !== null) return;
    termination = terminateProcessTree(child).catch((error: unknown) => {
      cleanupError = error instanceof Error ? error.message : String(error);
      // A leader-only fallback cannot confirm descendant cleanup, so keep the error.
      try { child.kill("SIGKILL"); } catch { /* Report the cleanup failure below. */ }
    }).finally(() => {
      terminationCompletedAt = performance.now();
    });
  };

  const watchdog = setInterval(() => {
    const now = performance.now();
    if (terminalStatus === null) {
      if (publicationError !== null) {
        terminalStatus = "failed";
        terminate();
      } else if (execution.isCancelled()) {
        terminalStatus = "cancelled";
        terminate();
      } else if (options.timeout >= 0 && now - startedAt >= timeoutMs) {
        terminalStatus = "timed_out";
        terminate();
      }
    }
    // Descendants can inherit pipes after the shell exits. Keep active output,
    // but release quiet handles rather than waiting indefinitely for EOF.
    if (exited && now - lastOutputAt >= POST_EXIT_IDLE_MS) closeOutput();
    if (terminationCompletedAt !== null && now - terminationCompletedAt >= TERMINATION_EXIT_GRACE_MS) {
      closeOutput(true);
      if (!exited) {
        cleanupError ??= "Bash did not exit after process termination";
        child.unref();
        resolveExit({ code: null, signal: null });
      }
    }
    if (now - lastFlush >= FLUSH_INTERVAL_MS) {
      flush();
      lastFlush = now;
    }
  }, WATCHDOG_INTERVAL_MS);
  watchdog.unref();

  try {
    await Promise.all([stdoutDone, stderrDone, exitInfo]);
  } finally {
    clearInterval(watchdog);
    child.removeListener("exit", onExit);
    await termination;
    await output.close();
  }
  await started;
  await publishing;
  flush();
  await publishing;

  const { code, signal } = await exitInfo;
  const exitCode = toExitCode(code, signal);
  const snapshot = output.snapshot();
  const stdout = snapshot.stdout.text;
  const stderr = snapshot.stderr.text;
  const durationMs = Math.trunc(performance.now() - invokedAt);
  const truncated = snapshot.stdout.truncated || snapshot.stderr.truncated;

  let result: BashResult;
  if (terminalStatus === "cancelled") {
    result = new BashResult({
      status: "cancelled",
      stdout,
      stderr,
      exitCode,
      error: execution.cancellationReason,
      durationMs,
      truncated,
    });
  } else if (terminalStatus === "timed_out") {
    result = new BashResult({
      status: "timed_out",
      stdout,
      stderr,
      exitCode,
      error: `Bash command timed out after ${options.timeout} seconds`,
      durationMs,
      truncated,
    });
  } else if (exitCode === 0) {
    result = new BashResult({
      status: "completed",
      stdout,
      stderr,
      exitCode,
      durationMs,
      truncated,
    });
  } else {
    result = new BashResult({
      status: "failed",
      stdout,
      stderr,
      exitCode,
      error: `Bash command exited with code ${exitCode}`,
      durationMs,
      truncated,
    });
  }

  if (cleanupError !== null) {
    result.error = `${result.error ?? "Bash execution failed"}; Process cleanup failed: ${cleanupError}`;
  }
  result.outputComplete = !outputIncomplete;
  result.outputFiles = output.files;
  result.outputFileComplete = output.files !== null && output.fileError === null && !outputIncomplete;
  result.outputFileError = output.fileError;
  result.stdoutStartMidLine = snapshot.stdout.startMidLine;
  result.stderrStartMidLine = snapshot.stderr.startMidLine;
  if (publicationError !== null) {
    result.status = "failed";
    result.error = publicationError + (cleanupError === null ? "" : `; Process cleanup failed: ${cleanupError}`);
  }
  return await finish(result);
}
